### PR TITLE
add base Bucket contract

### PR DESCRIPTION
**Pull request by gravityblast**
Add a base Bucket contract from which GiftBucket and NFTBucket inherit.

### CODE REVIEW

**Comment by gravityblast** [comment]
@bitgamma since a lot of things moved around I think it would be easier to review the final contract files instead of the changes.

File: contracts/Bucket.sol [hunk 1]
```diff
@@ -1,15 +1,91 @@
...
+  uint256 public startTime;
+
+  uint256 constant maxTxDelayInBlocks = 10;
+  bytes32 constant REDEEM_TYPEHASH = keccak256("Redeem(uint256 blockNumber,bytes32 blockHash,address receiver,bytes32 code)");
+  bytes32 constant EIP712DOMAIN_TYPEHASH = keccak256("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)");
+  bytes32 DOMAIN_SEPARATOR;
+
+  struct Gift {
+    address recipient;
+    bytes32 code;
```
Comment (bitgamma) [inline contracts/Bucket.sol:18]: can we leave this uint256?

Comment (gravityblast) [reply contracts/Bucket.sol:18]: the redeem code has always been bytes32, would you prefer to change it to uint256?

Comment (bitgamma) [reply contracts/Bucket.sol:18]: sorry my error :D

**Pull request merged**